### PR TITLE
Fix dlpack type signatures to match Array API spec.

### LINKED ISSUE #17510: Support dlpack as per array API spec
JAX already supports DLPack via `jax.dlpack.to_dlpack` and `jax.dlpack.from_dlpack`. If `jax.Array` were compatible with the Protocol below we'd have a much easier time to interop between array API compatible frameworks:

```python
from jax import numpy as jnp
from typing import Protocol, Any
from enum import Enum

PyCapsule = Any

class SupportsDLPack(Protocol):
    def __dlpack__(self, *, stream: int | Any | None = None) -> PyCapsule:
        """Export the `Array` as DLPack capsule.

        Args:
            stream:
                For CUDA and ROCm a Python integer representing a pointer to a
                stream, on devices that support streams.

        Returns:
            A DLPack capsule for the `Array`.

        Raises:
            BufferError:
                If the data cannot be exported as DLPack.
        """
        ...

    def __dlpack_device__(self) -> tuple[Enum, int]:
        """The device type and device ID in DLPack format.

        Returns:
            A tuple `(device_type, device_id)` in DLPack format.
        """
        ...
        
def f(x: SupportsDLPack) -> None:
    return None

f(jnp.ones(1))  # pyright: ignore [reportUnknownMemberType]
```

```
[Pyright] Argument of type "Array" cannot be assigned to parameter "x" of type "SupportsDLPack" in function "f"
  "Array" is incompatible with protocol "SupportsDLPack"
    "__dlpack_device__" is not present
      "__dlpack__" is an incompatible type
        Type "() -> Any" cannot be assigned to type "(*, stream: int | Any | None = None) -> PyCapsule"
          Keyword parameter "stream" is missing in source
```

There is already ongoing work for array API compatibility at https://github.com/google/jax/pull/16099. This issue is just a minor change required to the annotations in the dlpack implementation in `jax.Array` tough (and an "optional keyword" -> "keyword only" change).

`dlpack` in the the API spec: 
- https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
- https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack_device__.html

cc @jakevdp

### PR DESCRIPTION
**Pull request by hawkinsp**
Fixes https://github.com/google/jax/issues/17510

### CODE REVIEW

File: jax/_src/array.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 import operator as op
 import numpy as np
 import functools
-from typing import Any, Callable, cast, TYPE_CHECKING
+from typing import Any, Enum, Callable, cast, TYPE_CHECKING
```
Comment (aaronmondal) [inline jax/_src/array.py:22]: This needs to come from the `enum` module:

```python
from enum import Enum
```

**Review by aaronmondal — CHANGES_REQUESTED** [review]

File: jax/_src/array.py (hunk 1 above)
Comment (hawkinsp) [reply jax/_src/array.py]: Done.

**Review by jakevdp — APPROVED** [review]
Looks good!

I wonder if we should add code somewhere that serves to test these annotations? I don't think we have any coverage of that, because we skip most test files in both mypy and pytype.

**Review by aaronmondal — APPROVED** [review]

**Comment by hawkinsp** [comment]
> Looks good!
> 
> I wonder if we should add code somewhere that serves to test these annotations? I don't think we have any coverage of that, because we skip most test files in both mypy and pytype.

Yeah, that's a general problem with all of our API types, not specific to dlpack.

I think the best fix would be to just apply type checkers to our tests at some point.

**Comment by aaronmondal** [comment]
I've also been looking for a way to test things like this somehow. Instance checks etc don't work (or at least are tricky to use with all the generic stuff etc) we've settled for using typechecking in pre-commit.

The way we're testing at the moment is using pyright (though mypy should achieve the same) in pre-commit and using the protocol below. Note that it's not 100% accurate yet and still needs some tweaks. We're then annotating everything with that so that typecheckers crash when something is not according to spec.

<details>
  <summary>Protocol</summary>

```python
"""Protocol to check for adherance to the Array API.

- [The API spec](https://data-apis.org/array-api/2022.12/API_specification/index.html)
- [Support in Numpy](https://numpy.org/devdocs/reference/array_api.html)
- [Upcoming support in JAX](https://github.com/google/jax/pull/16099)
"""

from enum import Enum
from types import EllipsisType
from typing import Any, Protocol, Self, TypeVar, runtime_checkable

from numpy.typing import DTypeLike

ellipsis = EllipsisType
DType_co = TypeVar('DType_co', covariant=True, bound=DTypeLike)
PyCapsule = Any
Device = Any

# TODO(aaron): Bug that covariant DType_co is flagged as invalid?
@runtime_checkable
class Array(Protocol[DType_co]): # pyright: ignore [reportInvalidTypeVarUse]
    """An array structure that is compliant with the Array API."""

    # Properties

    @property
    def dtype(self) -> DType_co:
        """Data type of the array elements."""
        ...

    @property
    def device(self) -> Device:
        """Hardware device the array data resides on."""
        ...

    @property
    def mT(self) -> Self:  # noqa: N802
        """Transpose of a matrix or stack of matrices."""
        ...

    @property
    def ndim(self) -> int:
        """The number of `Array` dimensions (axes)."""
        ...

    @property
    def shape(self) -> tuple[int | None, ...]:
        """Array dimensions."""
        ...

    @property
    def size(self) -> int | None:
        """The number of elements in the `Array`."""
        ...

    @property
    def T(self) -> Self:  # noqa: N802
        """The transposed array.

        Only defined on two-dimensional `Array`. Raises an error otherwise.
        """
        ...

    # Arithmetic operators

    def __pos__(self) -> Self:
        """Evaluate `+self_i` for each element of an array instance.

        Returns:
            An `Array` containing the evaluated result for each element. The
            returned array must have the same data type as `self`.
        """
        ...

    def __neg__(self) -> Self:
        """Evaluate `-self_i` for each element of an array instance.

        Returns:
            An `Array` containing the evaluated result for each element in
            `self`.
        """
        ...

    def __add__(self, other: float | Self, /) -> Self:
        """Element-wise sum with `other`.

        Args:
            other:
                `Array` instance. Must be compatible with `self`. Should have
                numeric data type.

        Returns:
            An `Array` containing the element-wise sums.
        """
        ...

    def __sub__(self, other: float | Self, /) -> Self:
        """Element-wise difference of an `Array` instance.

        Args:
            other:
                Subtrahend `Array`.

        Returns:
            An `Array` containing the element-wise differences.
        """
        ...

    def __mul__(self, other: float | Self, /) -> Self:
        """Element-wise product of the `Array` with `other`.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise products.
        """
        ...

    def __truediv__(self, other: float | Self, /) -> Self:
        """Evaluate `self_i / other_i` element-wise.

        Args:
            other:
                The divisor `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __floordiv__(self, other: float | Self, /) -> Self:
        """Evaluate `self_i // other_i` element-wise.

        Args:
            other:
                The divisor `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __mod__(self, other: float | Self, /) -> Self:
        """Evaluate `self_i % other_i` element-wise.

        Args:
            other:
                The modulus `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __pow__(self, other: float | Self, /) -> Self:
        """Element-wise potentiation `self_i ** other_i`.

        Args:
            other:
                The exponent `Array`.

        Returns:
            An `Array` containing the element-wise-results.
        """
        ...

    # Array operators

    def __matmul__(self, other: Self, /) -> Self:
        """Compute the matrix product.

        TODO(aaron): Describe this in detail.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array`.
        """
        ...

    # Bitwise operators

    def __invert__(self) -> Self:
        """Evaluate `~self_i` element-wise.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __and__(self, other: int | bool | Self, /) -> Self:
        """Evaluate `self_i & other_i` element-wise.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __or__(self, other: int | bool | Self, /) -> Self:
        """Evaluate `self_i | other_i` element-wise.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __xor__(self, other: int | bool | Self, /) -> Self:
        """Evaluate `self_i ^ other_i` element-wise.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...


    def __lshift__(self, other: int | Self, /) -> Self:
        """Evaluate `self_i << other_i` element-wise.

        Args:
            other:
                An `Array`. Each element must be greater than or equal to `0`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __rshift__(self, other: int | Self, /) -> Self:
        """Evaluate `self_i >> other_i` element-wise.

        Args:
            other:
                An `Array`. Each element must be greater than or equal to `0`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    # Comparison operators

    def __lt__(self, other: float | Self, /) -> Self:
        """Compute the truth value of `self_i < other_i` element-wise.

        Args:
            other:
                A real-valued `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __le__(self, other: float | Self, /) -> Self:
        """Compute the truth value of `self_i <= other_i` element-wise.

        Args:
            other:
                A real-valued `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __gt__(self, other: float | Self, /) -> Self:
        """Compute the truth value of `self_i > other_i` element-wise.

        Args:
            other:
                A real-valued `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __ge__(self, other: float | Self, /) -> Self:
        """Compute the truth value of `self_i >= other_i` element-wise.

        Args:
            other:
                A real-valued `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __eq__( # pyright: ignore [reportIncompatibleMethodOverride]
        self,
        other: float | bool | Self,
        /,
    ) -> Self:
        """Compute the truth value of `self_i == other_i` element-wise.

        Args:
            other:
                A real-valued `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __ne__( # pyright: ignore [reportIncompatibleMethodOverride]
        self,
        other: float | bool | Self,
        /,
    ) -> Self:
        """Compute the truth value of `self_i != other_i` element-wise.

        Args:
            other:
                A real-valued `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    # In-place operators

    def __iadd__(self, other: Self, /) -> Self:
        """In-place addition.

        Args:
            other:
                An `Array`.
        """
        ...

    def __isub__(self, other: Self, /) -> Self:
        """In-place subtraction.

        Args:
            other:
                An `Array`.
        """
        ...

    def __imul__(self, other: Self, /) -> Self:
        """In-place multiplication.

        Args:
            other:
                An `Array`.
        """
        ...

    def __itruediv__(self, other: Self, /) -> Self:
        """In-place division.

        Args:
            other:
                An `Array`.
        """
        ...

    def __ifloordiv__(self, other: Self, /) -> Self:
        """In-place floored division.

        Args:
            other:
                An `Array`.
        """
        ...

    def __ipow__(self, other: Self, /) -> Self:
        """In-place exponentiation.

        Args:
            other:
                An `Array`.
        """
        ...

    def __imod__(self, other: Self, /) -> Self:
        """In-place modulus.

        Args:
            other:
                An `Array`.
        """
        ...

    def __imatmul__(self, other: Self, /) -> Self:
        """In-place Matrix multiplication.

        Args:
            other:
                An `Array`.
        """
        ...

    def __iand__(self, other: Self, /) -> Self:
        """In-place `and`.

        Args:
            other:
                An `Array`.
        """
        ...

    def __ior__(self, other: Self, /) -> Self:
        """In-place `or`.

        Args:
            other:
                An `Array`.
        """
        ...

    def __ixor__(self, other: Self, /) -> Self:
        """In-place `xor`.

        Args:
            other:
                An `Array`.
        """
        ...

    def __ilshift__(self, other: Self, /) -> Self:
        """In-place `<<`.

        Args:
            other:
                An `Array`.
        """
        ...

    def __irshift__(self, other: Self, /) -> Self:
        """In-place `>>`.

        Args:
            other:
                An `Array`.
        """
        ...

    # Reflected operators

    def __radd__(self, other: float | Self, /) -> Self:
        """Reflected element-wise sum with `other`.

        Args:
            other:
                `Array` instance. Must be compatible with `self`. Should have
                numeric data type.

        Returns:
            An `Array` containing the element-wise sums.
        """
        ...

    def __rsub__(self, other: float | Self, /) -> Self:
        """Reflected element-wise difference of an `Array` instance.

        Args:
            other:
                Subtrahend `Array`.

        Returns:
            An `Array` containing the element-wise differences.
        """
        ...

    def __rmul__(self, other: float | Self, /) -> Self:
        """Reflected element-wise product of the `Array` with `other`.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise products.
        """
        ...

    def __rtruediv__(self, other: float | Self, /) -> Self:
        """Reflected element-wise division.

        Args:
            other:
                The divisor `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __rfloordiv__(self, other: float | Self, /) -> Self:
        """Reflected element-wise division with rest.

        Args:
            other:
                The divisor `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __rpow__(self, other: float | Self, /) -> Self:
        """Reflected element-wise potentiation.

        Args:
            other:
                The exponent `Array`.

        Returns:
            An `Array` containing the element-wise-results.
        """
        ...

    def __rmod__(self, other: float | Self, /) -> Self:
        """Reflected element-wise modulus.

        Args:
            other:
                The modulus `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...


    def __rmatmul__(self, other: Self, /) -> Self:
        """Reflected matrix product.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array`.
        """
        ...

    def __rand__(self, other: int | bool | Self, /) -> Self:
        """Reflected element-wise `&` operation.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __ror__(self, other: int | bool | Self, /) -> Self:
        """Reflected element-wise `|` operation.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __rxor__(self, other: int | bool | Self, /) -> Self:
        """Reflected element-wise `^` operation.

        Args:
            other:
                An `Array`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...


    def __rlshift__(self, other: int | Self, /) -> Self:
        """Reflected element-wise `<<` operation.

        Args:
            other:
                An `Array`. Each element must be greater than or equal to `0`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    def __rrshift__(self, other: int | Self, /) -> Self:
        """Reflected element-wise `>>` operation.

        Args:
            other:
                An `Array`. Each element must be greater than or equal to `0`.

        Returns:
            An `Array` containing the element-wise results.
        """
        ...

    # Other methods

    def __abs__(self) -> Self:
        """Calculate the absolute value element-wise.

        Returns:
            An `Array`.
        """
        ...

    def __array_namespace__(*, api_version: str | None = None) -> Any:
        """Return an object that has all array API functions on it.

        Args:
            api_verions: String representing the version of the array API
            specification to be returned.
        """
        ...

    def __bool__(self) -> bool:
        """Convert a zero-dimensional `Array` to a Python `bool` object.

        Returns:
            A `bool` representing the single element of the `Array`.

        Notes:
            For real-valued floating-point operands:

            - If `self` is `NaN` the result is `True`.
            - If `self` is `+infinity` or `-infinity` the result is `True`.
            - If `self` is `+0` or `-0` the result is `False`
        """
        ...

    def __complex__(self) -> complex:
        """Convert a zero-dimensional `Array` to a Python `complex` object.

        Returns:
            A Python `complex` object representing the single element of the
            `Array` instance.

        Notes:
            For boolean operands:

            - If `self` is `True` the result is `1+0j`.
            - If `self` is `False` the result is `0+0j`.

            For real-valued floating-point operands:

            - If `self` is `Nan` the result is `NaN + NaN j`.
            - If `self` is `+infinity` the result is `+infinity + 0j`.
            - If `self` is `-infinity` the result is `-infinity + 0j`.
            - If `self` is a finite number the result is `self + 0j`.
        """
        ...

    # TODO(aaron): Should be `stream: int | Any | None = None`. Currently
    #              incompatible with numpy though.
    def __dlpack__(*, stream:  Any = None) -> PyCapsule:
        """Export the `Array` as DLPack capsule.

        Args:
            stream:
                For CUDA and ROCm a Python integer representing a pointer to a
                stream, on devices that support streams.

        Returns:
            A DLPack capsule for the `Array`.

        Raises:
            BufferError:
                If the data cannot be exported as DLPack.
        """
        ...

    def __dlpack_device__(self) -> tuple[Enum, int]:
        """The device type and device ID in DLPack format.

        Returns:
            A tuple `(device_type, device_id)` in DLPack format.
        """
        ...

    def __float__(self) -> float:
        """Convert a zero-dimensional `Array` to a Python `float` object.

        Returns:
            A Python `float` object representing the single element of the
            `Array` instance.

        Notes:
            For boolean operands:

            - If `self` is `True` the result is `1`.
            - If `self` is `False` the result is `0`.
        """
        ...

    def __getitem__(
        self,
        key: int
        | slice
        | ellipsis
        | tuple[int | slice | ellipsis, ...]
        | Self,
        /,
    ) -> Self:
        """Return `self[key]`.

        Args:
            key:
                The index key.

        Returns:
            An `Array` containing the accessed values.
        """
        ...

    def __index__(self) -> int:
        """Convert a zero-dimensional integer array to a Python `int` object.

        Returns:
            A Python `int` object representing the single element of the `Array`
            instance.
        """
        ...

    def __setitem__(
        self,
        key: int
        | slice
        | ellipsis
        | tuple[int | slice | ellipsis, ...]
        | Self,
        value: float | bool | Self,
        /,
    ) -> None:
        """Set `self[key]` to `value`.

        Args:
            key:
                The index key.
            value:
                The value(s) to set.
        """
        ...

    def to_device(
        self,
        device: Device,
        /,
        *,
        stream: None = None,  # TODO(aaron): stream: int | Any | None = None
    ) -> Self:
        """Copy the `Array` from it's current device to the specified `device`.

        Args:
            device:
                A `device` object.
            stream:
                A stream object to use during the copy.
        """
        ...
```

<details/>

**Pull request merged**